streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableTransformValues.java

```diff
@@ -0,0 +1,256 @@
+
+        final OldValueGetter<K, V, V1> oldValueGetter;
+        if (!sendOldValues) {
+            oldValueGetter = new NullOldValueGetter<>();
+        } else if (queryableName != null) {
+            oldValueGetter = new MaterializedOldValueGetter<>(getMaterializedValueGetterSupplier());
+        } else {
+            oldValueGetter = new TransformingOldValueGetter<>(valueTransformer);
+        }
+
+        return new KTableTransformValuesProcessor(valueTransformer, oldValueGetter);
+    }
+
+    @Override
+    public KTableValueGetterSupplier<K, V1> view() {
+        if (queryableName != null) {
+            return getMaterializedValueGetterSupplier();
+        }
+
+        return new KTableValueGetterSupplier<K, V1>() {
+            final KTableValueGetterSupplier<K, V> parentValueGetterSupplier = parent.valueGetterSupplier();
+
+            public KTableValueGetter<K, V1> get() {
+                return new KTableTransformValuesGetter(
+                        parentValueGetterSupplier.get(),
+                        transformerSupplier.get());
+            }
+
+            @Override
+            public String[] storeNames() {
+                return parentValueGetterSupplier.storeNames();
+            }
+        };
+    }
+
+    @Override
+    public void enableSendingOldValues() {
+        parent.enableSendingOldValues();
+        sendOldValues = true;
+    }
+
+    private KTableValueGetterSupplier<K, V1> getMaterializedValueGetterSupplier() {
+        return new KTableMaterializedValueGetterSupplier<>(queryableName);
+    }
+
+    private static <K, V, V1> V1 computeValue(final K key,
+                                              final V value,
+                                              final ValueTransformerWithKey<? super K, ? super V, ? extends V1> valueTransformer) {
+        if (value == null) {
+            return null;
+        }
+
+        return valueTransformer.transform(key, value);
+    }
+
+    private class KTableTransformValuesProcessor extends AbstractProcessor<K, Change<V>> {
+        private final ValueTransformerWithKey<? super K, ? super V, ? extends V1> valueTransformer;
+        private final OldValueGetter<K, V, V1> oldValueGetter;
+        private KeyValueStore<K, V1> store;
+        private TupleForwarder<K, V1> tupleForwarder;
+
+        private KTableTransformValuesProcessor(final ValueTransformerWithKey<? super K, ? super V, ? extends V1> valueTransformer,
+                                               final OldValueGetter<K, V, V1> oldValueGetter) {
+            this.valueTransformer = Objects.requireNonNull(valueTransformer, "valueTransformer");
+            this.oldValueGetter = Objects.requireNonNull(oldValueGetter, "oldValueGetter");
+        }
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public void init(final ProcessorContext context) {
+            super.init(context);
+
+            oldValueGetter.init(context);
```

Comment:
This seems an overkill to me: for old values, we only need `change.oldValue`, and actually all of its implementations either do not need the value, or only look for `change.oldValue`.

We can probably simply it in line 132, just as:

```
oldValue = queryableName ? /* get old value from materialized store*/ : computeValue(key, change.oldValue, valueTransformer);
```

Reply:
Actually I think `using the materialized store to get the old value` can be a separate optimization PR by itself: we should consider doing this for filter, mapValues, transformValues, etc, all operators.

To keep the scope / loc of this PR small, we can consider only use `computeValue(key, change.oldValue, valueTransformer)`

Reply:
This does mean backing out changes that will make this PR _less_ correct. I'm happy to do it, but my opinion would be to leave it in.

With regards to the overkill...  yep, I think I know what you're saying - just get the store from the context, and retrieve the value - I'll sort that.